File: backend/app/schemas/team.py
"""
Team-related Pydantic schemas
"""

from datetime import datetime
from typing import Optional, List
from uuid import UUID
from enum import Enum

from pydantic import BaseModel, EmailStr, Field


class TeamRole(str, Enum):
    """Team member roles"""
    OWNER = "owner"
    ADMIN = "admin"
    MEMBER = "member"
    VIEWER = "viewer"


# ============ Team Schemas ============

class TeamCreate(BaseModel):
    """Schema for creating a new team"""
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    slug: str = Field(..., min_length=1, max_length=100, pattern=r"^[a-z0-9-]+$")


class TeamUpdate(BaseModel):
    """Schema for updating a team"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None


class TeamMemberResponse(BaseModel):
    """Team member response schema"""
    id: UUID
    user_id: UUID
    user_email: str
    user_name: str
    role: TeamRole
    joined_at: datetime

    class Config:
        from_attributes = True


class TeamResponse(BaseModel):
    """Team response schema"""
    id: UUID
    name: str
    description: Optional[str]
    slug: str
    owner_id: UUID
    is_active: bool
    created_at: datetime
    updated_at: datetime
    member_count: Optional[int] = None

    class Config:
        from_attributes = True


class TeamDetailResponse(TeamResponse):
    """Detailed team response with members"""
    members: List[TeamMemberResponse] = []


class TeamListResponse(BaseModel):
    """Paginated list of teams"""
    items: List[TeamResponse]
    total: int
    page: int
    size: int
    pages: int


# ============ Team Member Schemas ============

class TeamMemberAdd(BaseModel):
    """Schema for adding a member to team"""
    user_id: UUID
    role: TeamRole = TeamRole.MEMBER


class TeamMemberUpdate(BaseModel):
    """Schema for updating member role"""
    role: TeamRole


# ============ Team Invitation Schemas ============

class TeamInvitationCreate(BaseModel):
    """Schema for creating an invitation"""
    email: EmailStr
    role: TeamRole = TeamRole.MEMBER


class TeamInvitationResponse(BaseModel):
    """Team invitation response schema"""
    id: UUID
    team_id: UUID
    team_name: str
    email: str
    role: TeamRole
    invited_by_name: str
    expires_at: datetime
    created_at: datetime
    is_expired: bool

    class Config:
        from_attributes = True


class TeamInvitationAccept(BaseModel):
    """Schema for accepting an invitation"""
    token: str

File: frontend/types/team.ts
/**
 * Team Types
 */

export type TeamRole = 'owner' | 'admin' | 'member' | 'viewer';

export interface TeamMember {
  id: string;
  userId: string;
  userEmail: string;
  userName: string;
  role: TeamRole;
  joinedAt: string;
}

export interface Team {
  id: string;
  name: string;
  description?: string;
  slug: string;
  ownerId: string;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  memberCount?: number;
}

export interface TeamDetail extends Team {
  members: TeamMember[];
}

export interface TeamCreate {
  name: string;
  description?: string;
  slug: string;
}

export interface TeamUpdate {
  name?: string;
  description?: string;
}

export interface TeamListResponse {
  items: Team[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface TeamMemberAdd {
  userId: string;
  role: TeamRole;
}

export interface TeamMemberUpdate {
  role: TeamRole;
}

export interface TeamInvitation {
  id: string;
  teamId: string;
  teamName: string;
  email: string;
  role: TeamRole;
  invitedByName: string;
  expiresAt: string;
  createdAt: string;
  isExpired: boolean;
}

export interface TeamInvitationCreate {
  email: string;
  role?: TeamRole;
}

export interface TeamInvitationAccept {
  token: string;
}

export const ROLE_LABELS: Record<TeamRole, string> = {
  owner: '소유자',
  admin: '관리자',
  member: '멤버',
  viewer: '뷰어',
};

export const ROLE_DESCRIPTIONS: Record<TeamRole, string> = {
  owner: '팀의 모든 권한을 가집니다',
  admin: '멤버 관리 및 설정 변경 가능',
  member: '회의 생성 및 편집 가능',
  viewer: '회의 조회만 가능',
};

export function canManageTeam(role: TeamRole): boolean {
  return role === 'owner' || role === 'admin';
}

export function canEditMeeting(role: TeamRole): boolean {
  return role !== 'viewer';
}
